vassal-app/src/main/java/VASSAL/script/expression/AuditTrail.java

```diff
@@ -189,9 +197,30 @@ public void setSource(Auditable source) {
   public String toString() {
     final StringBuilder sb = new StringBuilder("Expression Audit:");
 
+    // Add some source identification to the audit trail
+    if (source != null) {
+      // Calling getComponentName() on a GamePiece may cause further expression evaluations, which could fail.
+      // Set the reportingInProgress flag on which will prevent the creation of any more audit trails until
+      // this trail has completed reporting to prevent an infinite recursive cycle of Audit Trail creation.
+      final boolean origReportingState = isReportingInProgress();
+      setReportingInProgress(true);
+      if (source instanceof EditablePiece) {
+        sb.append("\n   ").append(Resources.getString("Audit.source_type", "Piece Trait"));
+        sb.append("\n   ").append(Resources.getString("Audit.source_name", source.getComponentName()));
+        sb.append("\n   ").append(Resources.getString("Audit.source_description", source.getComponentTypeName()));
```

Comment:
You can chain them all, just drop the `sb` at the start of successive lines and remove the semicolons.

Reply:
Good point.